RGPosadas/LitterRally
# US-5.1: Waste Recognition Tool

closes #33

This PR implements the waste recognition camera feature.

How to test:
1. Start the UI and API applications,
2. navigate to `http://localhost:4200/camera` by using the camera icon
3. point the camera at stuff and see the detections appear!



Example:
![unknown](https://user-images.githubusercontent.com/23544999/109398234-a384fb80-7909-11eb-80aa-415956ab95b0.png)

Notes: 
* The model can only detect some objects, I will update the model soon, but for now, see [this file](https://github.com/RGPosadas/Mull/blob/develop/apps/mull-ui/src/app/services/maps.ts#L12) to understand what objects can be detected
* You can adjust the number of results and score threshold in `waste-recognition-page.tsx` if you're having trouble getting good detection, let me know if you find values that work well for you.
* If you want to test it on your phone, you can host the app on your machine's IP address and visit the IP address on your phone. However, keep in mind that you will need to mess around with your browser settings since camera access is denied by default on non-https connections that aren't localhost: https://stackoverflow.com/questions/34197653/getusermedia-in-chrome-47-without-using-https
* Also, renamed trash-recognition-page to waste-recognition-page
* For the unit test, some parts of the code weren't able to be fully tested and covered, since I needed to be able to trigger the video onloadedmetadata event, and I wasn't able to figure out how to do it in Jest
* For the e2e tests, to fully test the page functionality, I would need to set a video to be the mocked camera feed and see if detections go through as expected. However I wasn't able to get it to work, so I decided to not waste too much time on it and keep it simple
